packages/server/src/unstable-core-do-not-import/parser.ts

```diff
@@ -110,5 +118,25 @@ export function getParseFn<TType>(procedureParser: Parser): ParseFn<TType> {
     };
   }
 
+  if ('~standard' in parser) {
+    // StandardSchemaEsque
+    return async (value) => {
+      const result = await parser['~standard'].validate(value);
+      if (result.issues) {
+        throw new SchemaError(result.issues);
+      }
+      return result.value;
+    };
+  }
+
   throw new Error('Could not find a validator fn');
 }
+
+class SchemaError extends Error {
+  public readonly issues: ReadonlyArray<v1.StandardIssue>;
+  constructor(issues: ReadonlyArray<v1.StandardIssue>) {
+    super(issues[0]?.message);
+    this.name = 'SchemaError';
+    this.issues = issues;
+  }
+}
```

Comment:
Zod will probably still have inferrable errors via `.parse`, but throwing errors via the standard schema interface would force any schema library to add the spec as a runtime dependency to throw the same error class. This is probably not an option. If you prefer the previous implementation with `throw new Error('Schema Error:', { cause: result.issues });` I can revert it.

Reply:
Even without the error class `result.issues` could be generic, no?

Reply:
What would be the advantage of having `result.issues` as a generic for a library like tRPC?

Reply:
Right now, none.

I'm just thinking about the spec in general and want it to be as good as possible. When doing form validation it'd be nice to have the possible fields that might error be autocompletable for example

Reply:
I am not sure if `result.issues` as a generic will help with autocompletable fields, as this information can already be extracted from the input type of the schema using a [`Paths` utility type](https://github.com/ts-essentials/ts-essentials/tree/master/lib/paths). So far, the only schema library I know of that has almost type-safe issues is Valibot. Most other libraries just return a general issue interface.

Reply:
![CleanShot 2024-11-12 at 19 24 29@2x](https://github.com/user-attachments/assets/44aa3012-5ec3-4a80-b53d-cfc2b752e9b0)

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wF5wGZQiOAIgwgBNCAoCgYwgDtV5VqALAUxGTgF5MA6CACMAVm2owAFAG8KcOXEHIoALn6MowOgHMJASgoBffTXqM4UNqgCuAG3i9m7Tn1TJsbAApLUbaUarA2HASAIQW1nYuVtTUlqi6cDLy5pa2MHxsUHhQANxyAPT5cAA8ALSlcFpsdJnA1LLyhclyAHoA-FTJtAzw2MBsNqQAolnQ6LzhaRmjUG182DbIMDDVevP9gyPZqA1yfQPDM6h8cE3IVjAQcLTgNmwwyIJ3u6dF7YZAA

Reply:
Thank you for this example. I understand you much better now. Since the Standard Schema spec is intentionally minimal, it most likely won't include a `flatten` method as part of its issues, but the spec already provides all the information to implement a `flatten` function yourself (or as part of our official [utils package](https://github.com/standard-schema/standard-schema/tree/main/packages/utils)).

A `flatten` function/method requires 3 things:
1. The **path information** to create a dot path
2. The **error messages** related to a specific path
3. The **input type** to make flatten type safe

Every issue of the Standard Schema spec includes a `.message` and optionally a `.path` property. This information can be used at runtime to create a flat error object. The input type can be used to make the output of `flatten` typesafe by extracting every possible nested path.

Reply:
To summarise, if I understood you correctly, you were basically asking whether we should additionally include the type information in `result.issues` to make it easier to extract something like `fieldErrros` in a typesafe way.

Reply:
alrght, sounds good.

then all that's left is to decide whether to use this custom `SchemaError` or `new Error("Schema validation failed", { cause: issues })`. What do you think? Will you try and recommend one path over the other for integrations?

Reply:
I think something like a `SchemaError` might make sense. On the other hand, we are still figuring out if we should provide a `@standard-schema/utils` package with runtime code like a `SchemaError` class or not.

In the long run, a goal would be to have tRPC and other similar libraries mainly or only support Standard Schema as a common interface for schema libraries. If these libraries also commit to a unified implementation, e.g. by using `SchemaError`, the DX for both sides will be great. But we are still in the phase of figuring things out. Feel free to share your thoughts!

The current plan is to add more libraries to Standard Schema by creating more such pull requests. This way we will hopefully get more feedback. In 2 or 3 weeks we could imagine releasing a first v1 RC of the spec and making Standard Schema more public, for example by talking more about it on X.